cql3/restrictions/restriction.hh

```diff
@@ -58,6 +59,55 @@ namespace restrictions {
 struct allow_local_index_tag {};
 using allow_local_index = bool_class<allow_local_index_tag>;
 
+/// Work-in-progress new restriction representation.  When complete, it will replace the old one.
+namespace wip {
+
+// The new representation exposes its member data publicly.  Operations on restrictions are
+// performed by free functions that take restrictions as parameters and use the visitor pattern to
+// specialize code for different kinds of restrictions.
+//
+// The most interesting class is wip::binary_operator below, which can represent both multi- and
+// single-column restrictions.  Instead of the old merge_with mechanism for combining restrictions,
+// we will simply add them into a conjunction expression that will be processed using visitors.
+
+class binary_operator;
+class conjunction;
+
+/// A restriction expression -- union of all possible restriction types.  bool means a Boolean constant.
+using expression = std::variant<bool, conjunction, binary_operator>;
+
+/// A column, optionally subscripted by a term (eg, c1 or c2['abc']).
+struct column_value {
+    const column_definition* col;
+    ::shared_ptr<term> sub; ///< If present, this LHS is col[sub], otherwise just col.
+    /// For easy creation of vector<column_value> from vector<column_definition*>.
+    column_value(const column_definition* col) : col(col) {}
+    /// The compiler doesn't auto-generate this due to the other constructor's existence.
+    column_value(const column_definition* col, ::shared_ptr<term> sub) : col(col), sub(sub) {}
```

Comment:
You don't need either constructor, see https://godbolt.org/z/bYKfvz. Plus, std::move(sub).

Reply:
I need both constructors for reasons explained in the blurbs.  See https://godbolt.org/z/WZ7bAf

Reply:
ok.